copt/utils.py

```diff
@@ -77,6 +77,45 @@ def safe_sparse_add(a, b):
                 b = b.ravel()
         return a + b
 
+@njit(nogil=True)
+def fast_csr_vm(x, data, indptr, indices, d, idx):
+    """
+
+    :param x:
+    :param data:
+    :param indptr:
+    :param indices:
+    :param d:
+    :param idx:
+    :return:
+    """
+    res = np.zeros(d)
+    assert x.shape[0] == len(idx)
+    for k, i in np.ndenumerate(idx):
+        for j in range(indptr[i], indptr[i+1]):
+            j_idx = indices[j]
+            res[j_idx] += x[k] * data[j]
+    return res
+
+
+@njit(nogil=True)
```

Comment:
Are both functions needed? It looks to me that one is a more general function that the other

Reply:
I think they are: they're both matrix-vector multiplications, but one is vM, the other is Mv.

Reply:
If we double-transpose, we get `vM = (M^Tv^T)^T`, but then `v^T` is a row, and not a column.